utils/src/main/java/software/amazon/awssdk/utils/async/DelegatingBufferingSubscriber.java

```diff
@@ -0,0 +1,144 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.utils.async;
+
+import static software.amazon.awssdk.utils.async.StoringSubscriber.EventType.ON_NEXT;
+
+import java.nio.ByteBuffer;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicLong;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+
+/**
+ * This Subscriber buffers the {@link ByteBuffer} it receives until the specified maximumBuffer is reached, at which point
+ * it sends those bytes to the delegate {@link Subscriber}.
+ */
+@SdkInternalApi
+public class DelegatingBufferingSubscriber extends DelegatingSubscriber<ByteBuffer, ByteBuffer> {
+
+    /**
+     * The maximum amount of bytes allowed to be stored in the StoringSubscriber
+     */
+    private final int maximumBuffer;
+
+    /**
+     * Current amount of bytes buffered in the StoringSubscriber
+     */
+    private final AtomicLong currentlyBuffered = new AtomicLong(0);
+
+    /**
+     * Stores the bytes received from the upstream publisher, awaiting sending them to the delegate once the buffer size is
+     * reached.
+     */
+    private final StoringSubscriber<ByteBuffer> storage = new StoringSubscriber<>(Integer.MAX_VALUE);
+
+    /**
+     * Publisher to send the ByteBuffer received from upstream publisher to the StoringSubscriber
+     */
+    private final SimplePublisher<ByteBuffer> publisherToStorage = new SimplePublisher<>();
+
+    /**
+     * Subscription of the publisher this subscriber subscribe to
+     */
+    private Subscription subscription;
+
+    public DelegatingBufferingSubscriber(int maximumBuffer, Subscriber<? super ByteBuffer> subscriber) {
+        super(subscriber);
+        this.maximumBuffer = maximumBuffer;
+    }
+
+    @Override
+    public void onSubscribe(Subscription subscription) {
+        // §2.05 must call subscription.cancel() if it already has a subscription and receives another on subscribe signal
+        if (this.subscription != null) {
+            subscription.cancel();
+            return;
+        }
+        super.onSubscribe(subscription);
+        this.subscription = subscription;
+        publisherToStorage.subscribe(storage);
+        subscription.request(maximumBuffer);
+    }
+
+    @Override
+    public void onNext(ByteBuffer byteBuffer) {
+        while (bufferOverflows(byteBuffer)) {
+            flushStorageToDelegate();
```

Comment:
We need to track downstream subscription; if there's no demand from downstream, we shouldn't call flush storage to delegate

Example: https://github.com/aws/aws-sdk-java-v2/blob/master/utils/src/main/java/software/amazon/awssdk/utils/async/FlatteningSubscriber.java

It's going to be super complex and hard to maintain if we add that logic... Let me see if we can reuse some of the logic in some way.

Reply:
Right, let me know if you think of something better

Reply:
let me know if you think of a better approach

Reply:
Well, can't think of another way, we just need to implement it :(